e2e: Don't try to create a UDP LoadBalancer on AWS

justinsb opened the pull request:
AWS doesn't support type=LoadBalancer with UDP services.  For now, we
simply skip over the test with type=LoadBalancer on AWS for the UDP
service.

Fix #20911

justinsb committed:
e2e: Don't try to create a UDP LoadBalancer on AWS

AWS doesn't support type=LoadBalancer with UDP services.  For now, we
simply skip over the test with type=LoadBalancer on AWS for the UDP
service.

Fix #20911

justinsb commented:
Still running these tests against AWS... please don't LGTM just yet!

ixdy commented:
cc @thockin

thockin commented:
LGTM.  Justin, feel free to self-lgtm when you think it's done.  let me know if you want another round of review for some reason :)

justinsb committed:
e2e: Allow longer for AWS LoadBalancers to start serving traffic

When we create a LoadBalancer on AWS, there is a longer delay after
creating the LB before it starts to serve traffic than there is on GCE.
A delay of a few minutes is normal.

Use a longer timeout when waiting for the LB on AWS therefore.

justinsb commented:
Further AWS e2e testing revealed that I had to add two more conditionals for some cases I missed (where it verified the previous state).  But passing on AWS now.

Also we were using the wrong timeout after the initial creation of the TCP LB (LagTimeout vs CreateTimeout); I also added a note about the fact that the create timeout is conflating two concepts: https://github.com/kubernetes/kubernetes/pull/20959/files#diff-20a4e2095b63ecd60dd25e78bcd67372R55  It only surfaces on AWS because there's a delay after the LB is created and appears in the API before it actually starts working (DNS propagation, I think).

Given all that, would you mind giving it a quick once-over again @thockin ?  I think it's fine, but don't feel entirely right doing a self-LGTM given that the extra changes aren't 100% trivial.  95% trivial, but not 100%...

thockin commented on test/e2e/service.go:
```diff
@@ -52,6 +52,10 @@ const loadBalancerLagTimeout = 2 * time.Minute
 
 // How long to wait for a load balancer to be created/modified.
 //TODO: once support ticket 21807001 is resolved, reduce this timeout back to something reasonable
+//TODO: this timeout is actually used in two distinct senses; once when waiting
```
This is not quite right - I used it again after changing the port number because that operation (on GCE) causes a full rebuild of the LB, so is akin to a create operation.

thockin commented on test/e2e/service.go:
```diff
@@ -567,10 +577,12 @@ var _ = Describe("Services", func() {
 
 		By("hitting the TCP service's LoadBalancer")
-		jig.TestReachableHTTP(tcpIngressIP, svcPort, loadBalancerLagTimeout)
+		jig.TestReachableHTTP(tcpIngressIP, svcPort, loadBalancerCreateTimeout)
```
this should be LagTimeout - we've already proven that the API says there is a LB.

justinsb replied:
Ah, my motivation here was that AWS was timing out waiting for the LB to be ready; it would not be surprising if the ELB took 5 minutes to start serving traffic after it was created in the API.  It might be we simply have to bump up LagTimeout on AWS.

I'll clean this up...

thockin commented on test/e2e/service.go:
```diff
@@ -567,10 +577,12 @@ var _ = Describe("Services", func() {
-		jig.TestReachableUDP(udpIngressIP, svcPort, loadBalancerLagTimeout)
+		if loadBalancerSupportsUDP {
+			By("hitting the UDP service's LoadBalancer")
+			jig.TestReachableUDP(udpIngressIP, svcPort, loadBalancerCreateTimeout)
```
LagTimeout

thockin commented:
only comments about the timeout value

justinsb commented:
Changed the timeout commit to just use a longer lag-timeout on AWS for load balancers (10 minutes).  You were right @thockin (of course) - the timeouts were set up correctly; just a little too short for the slow clouds :-)

thockin commented:
LGTM

justinsb commented:
That'll teach me to get cute and try skipping the tests.  I missed a usage of the constant in `test/e2e/cluster_upgrade.go`  Now we have some repeated code, but I think it's OK...

Sorry @thockin .  But now tested on GCE (automatically) and AWS (manually).

With this AWS has just one more e2e failure to go (`[It] should be able to up and down services`).  There's also a flakey test, but I'm hoping it also hits GCE and someone else will fix it before I get to it...

justinsb commented:
self-lgtm-ing, with thockin's approval on slack.

bprashanth commented:
@k8s-bot unit test this github issue: #21257

justinsb commented:
@k8s-bot test this github issue #21244

justinsb commented:
@k8s-bot test this github flake #21244

justinsb commented:
@k8s-bot unit test this github flake #21244

Not actually that flake - rather a weird error where there's just no unit test results - but we're already linked to it, so it shouldn't hurt.
